Adding option to hide unreleased exercises to course exercise list

niclasschuemann opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
Unreleased exercises are visible to tutors and instructors on the course exercise overview. Some tutors share their screen during tutor sessions and display the course exercise overview while navigating to the exercise they want to present. This leads to tutors "leaking" unreleased exercises on accident. To prevent this, tutors must be able to hide those exercises.
Adresses issues described in #2383 

### Description
Currently, it is possible to hide overdue exercises and exercises with full points. I have added a third checkbox that lets tutors and instructors hide unreleased exercises.

### Steps for Testing
You need: 1x Instructor 1x Tutor 1x Student 1x Test Course

1. Create a modeling exercise with a release date in the past
2. Create a modeling exercise and a quiz exercise with a release date in the future
3. As the student, navigate to the course exercise overview and make sure that
    - You **cannot** see the exercises with a release date in the future
    - You **cannot** see the checkbox "_Hide unreleased_"
4. As a tutor, navigate to the course exercise overview and make sure that
    - You **can** see the exercises with a release date in the future
    - You **can** see the checkbox "_Hide unreleased_"
    - The exercises with a release date in the future disappear if you check the checkbox and the exercise with a release date in 
      the past does not disappear
    - The exercises with a release date in the future appear again if you uncheck the checkbox 

### Screenshots
Without "Hide unreleased" enabled:
![num1](https://user-images.githubusercontent.com/45300855/115631815-da263380-a306-11eb-8a94-21ff97eb363e.JPG)

With "Hide unreleased" enabled:
![num2](https://user-images.githubusercontent.com/45300855/115631836-e3af9b80-a306-11eb-8077-6048f2367aad.JPG)

MaisiKoleni commented:
I am not sure if this should close #2383 already, see the comments that Stephan had in the issue. We either need a new issue or leave that one open because otherwise the suggestion of a collapsible "filter area" is lost / resolved too early.

JohannesStoehr reviewed (changes requested):
A quiz without a start time that wasn't set to visible is still shown after hiding all unreleased exercises.

![Bildschirmfoto 2021-04-22 um 08 49 53](https://user-images.githubusercontent.com/38322605/115668743-c5b65b00-a347-11eb-81df-656f60058e70.png)

niclasschuemann commented:
> I am not sure if this should close #2383 already, see the comments that Stephan had in the issue. We either need a new issue or leave that one open because otherwise the suggestion of a collapsible "filter area" is lost / resolved too early. I would recommend asking Stephan if we want this quick fix for the issue or refactor and tidy up the filter section as described in the comments (if you haven't done so already).

I agree, this PR does not cover all aspects described in the issue. Because EIST tutors are eagerly awaiting this enhancement I'd suggest we leave the scope of this PR as it is and take care of the other features later.

niclasschuemann commented:
> A quiz without a start time that wasn't set to visible is still shown after hiding all unreleased exercises.
> Apparently this quiz still has a release date internally.
> 
> ![Bildschirmfoto 2021-04-22 um 08 49 53](https://user-images.githubusercontent.com/38322605/115668743-c5b65b00-a347-11eb-81df-656f60058e70.png)

Weird, this case worked for me yesterday evening, but I will look into that again today evening. Thanks for testing!

krusche commented:
please make sure to only show this options to tutors and above (students don't need to see it)
Also make sure the option still looks nicely on small screens. I know that the additional options lead to overlapping elements which we should avoid.

niclasschuemann commented:
@JohannesStoehr I have fixed the issue you have found, thanks!

@krusche I made sure that this option is only visible to tutors and above. I changed the layout a little bit, so at least students won't have overlapping elements anymore:

**Student view:**
![student](https://user-images.githubusercontent.com/45300855/115802706-acf98400-a3df-11eb-93be-ff43f25c37a3.JPG)

**Tutor view:**
![tutor](https://user-images.githubusercontent.com/45300855/115802683-a539df80-a3df-11eb-9c00-9d4f5045de9e.JPG)

I acknowledge that this might not be the prettiest solution, but it should serve its purpose until we extend this functionality as described in issue [#2383](https://github.com/ls1intum/Artemis/issues/2383).

niclasschuemann commented:
@JohannesStoehr I have fixed the failing test and tested the changes again on TS1. It worked for me, so it would be great if you can review the changes as well :)

JohannesStoehr reviewed (dismissed):
The bug with the quiz is fixed.
But there seems to be a problem with some window widths and overlapping elements:
![Bildschirmfoto 2021-04-23 um 19 09 07](https://user-images.githubusercontent.com/38322605/115906131-6bacb700-a467-11eb-90d7-edc8fe3b9350.png)

niclasschuemann commented:
I have tried implementing the changes requested by @JohannesStoehr but I have not yet found a way to accomplish this. Help would be appreciated here or we fix this in a follow-up PR since this is not an issue introduced with this PR.

willergeorg reviewed (dismissed):
Tested on TS2. There I seem to have some issues with the checkbox feature. It shows up for an admin account and then worked perfectly fine for me. But for an account with only instructor role (TU1) the filter option does not show up in every course. Bizarrely, not all courses are accessible via the course management tap (don't know if this is on purpose and which logic this follows) and the filter option does only show for courses that are modifiable via course management for the specific account. But when it shows up, it works.

Edited and dismissed by @niclasschuemann:
The checkbox only appears in courses in which the user has instructor or tutor rights. As a server admin, the user has instructor rights in every course. If the user is not a server admin, the user might only have elevated permissions for specific courses. Only the courses in which the user has elevated permissions are displayed in the course management view. If the user sees a course in the course overview, but not in the course management view, the user is a student in this course.

NicolasUllmann reviewed (dismissed):
Tested on Server 1. I ran into the same problems as mentioned above with the checkbox. Somehow there seems to be a problem when using an instructor role.

Edited and dismissed by @niclasschuemann:
The checkbox only appears in courses in which the user has instructor or tutor rights. As a server admin, the user has instructor rights in every course. If the user is not a server admin, the user might only have elevated permissions for specific courses. Only the courses in which the user has elevated permissions are displayed in the course management view. If the user sees a course in the course overview, but not in the course management view, the user is a student in this course.

iozbeyli commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.html:
```diff
@@ -3,14 +3,20 @@
-                    <label class="d-flex align-items-center mb-0 mr-2">
-                        <input (change)="toggleFilters([filterType.NEEDS_WORK])" [checked]="activeFilters.has(filterType.NEEDS_WORK)" type="checkbox" />
-                        <span class="ml-2" jhiTranslate="artemisApp.courseOverview.exerciseList.fullScore">Hide full score</span>
+                    <label class="d-flex align-items-center mb-0 mr-2" *ngIf="course.isAtLeastTutor">
+                        <input (change)="toggleFilters([filterType.UNRELEASED])" [checked]="activeFilters.has(filterType.UNRELEASED)" type="checkbox" />
+                        <span class="ml-2" jhiTranslate="artemisApp.courseOverview.exerciseList.unreleased">Hide unreleased</span>
                     </label>
```
A small component for this can make everything cleaner, since it gets repeated 3 times in your changes.

niclasschuemann replied:
I agree with you, but we will address this later in a refactoring of the UI in this component

iozbeyli commented on src/test/javascript/spec/component/course/course-exercises.component.spec.ts:
```diff
@@ -182,7 +182,7 @@ describe('CourseExercisesComponent', () => {
 
         component.activeFilters.clear();
-        component.activeFilters.add(ExerciseFilter.OVERDUE);
+        component.activeFilters.add(ExerciseFilter.OVERDUE).add(ExerciseFilter.UNRELEASED);
```
You need to test how it affects the filtered result because you've changed that as well.

iozbeyli reviewed (dismissed):
Added a couple of comments on code. If you made a change that changes the behavior, you need to test that as well.

ArtemTsurikov reviewed (dismissed):
Tested on TS1: I have registered as artemis_test_user_1 (as a Student) and still was able to see the "Hide unreleased" checkbox and exercises with the release date in the future

katjanakosic reviewed (dismissed):
Tested on TS1. With the ArtemisTestUser2 and ArtemisTestUser6 everything worked as expected. When logged in as ArtemisTestUser1 I could still see the check box "hide unreleased" and the future exercises were displayed.

florian-glombik reviewed (changes requested):
There is an edge case with upcoming exercises, they are currently not hidden

![image](https://user-images.githubusercontent.com/63976129/116619687-12e68e00-a941-11eb-8b87-983260435bdd.png)

florian-glombik reviewed (approved):
Tested it on ts1 and worked as expected.
- students do not see the button
- tutors + instructors see the button and it hides the unreleased exercises (tested for modeling+quiz)
- tutors only see the button in courses where they are tutors
- the mentioned issue is fixed https://github.com/ls1intum/Artemis/pull/3266#pullrequestreview-648664619
